aptos-move/aptos-vm/src/parallel_executor/mod.rs

```diff
@@ -79,13 +85,31 @@ impl ParallelAptosVM {
         )
         .execute_transactions_parallel(state_view, signature_verified_block)
         {
-            Ok(results) => Ok((
-                results
-                    .into_iter()
-                    .map(AptosTransactionOutput::into)
-                    .collect(),
-                None,
-            )),
+            Ok(results) => {
+                // TODO: with more deltas, do this in parallel, and together with parallel
+                // execution's output processing. (Note: need to make DataCache trait
+                // to avoid circular dependencies).
+                let mut data_cache = StateViewCache::new(state_view);
+                Ok((
+                    results
+                        .into_iter()
+                        .map(|out| {
+                            let output_ext = AptosTransactionOutput::into(out);
+                            let (output, delta_writes) = output_ext
+                                .into_transaction_output(&data_cache)
+                                .expect("Delta application failed");
+
+                            if !output.status().is_discarded() {
+                                data_cache.push_write_set(
+                                    &delta_writes.expect("Expected materialized delta writeset"),
```

Comment:
do we need this? I remember data_cache is per block

Reply:
At the end of the parallel execution, currently, we still have output as a combination of writes and deltas, and need to transform the deltas into full writes. If we did it naively, resolving each delta could be a linear traversal (all the way to storage value or latest write), so instead we go bottom up and resolve it using data_cache, much like in sequential execution (but only for deltas). In other words, data cache must contains the resolved latest value for each path that previously had a delta.

There is an issue in this logic though, because we also have to make sure to push writes to the same paths that had a delta previously (currently it only pushes deltas). We test this case for PE but missed this stupid interconnect. I will fix it tonight.

Reply:
I will unresolve till I fix the issue.